### PR TITLE
[generic] fix attestation file creation when subject names are in subdirectories

### PR DESCRIPTION
**Pull request by asraa**
Signed-off-by: Asra Ali <asraa@google.com>

Updates https://github.com/slsa-framework/slsa-github-generator/issues/1225

The generic SLSA 3 generator attests to subjects given as input and their digest. However, if the subject is in a sub-directory, then it will fail creating the attestation because the sub-directory doesn't exist.

This retrieves the Base for the subject name, instead of the full path name.
@ianlewis I could also have created the sub-directory, but that made a lot less sense to me.

### CODE REVIEW

File: internal/builders/generic/attest.go [hunk 1]
```diff
@@ -59,7 +60,8 @@ run in the context of a Github Actions workflow.`,
 			// validated. This is done by CreateNewFileUnderCurrentDirectory.
 			if attPath == "" {
 				if len(parsedSubjects) == 1 {
-					attPath = fmt.Sprintf("%s.intoto.jsonl", parsedSubjects[0].Name)
+					filename := path.Base(parsedSubjects[0].Name)
```
Comment (ianlewis) [inline internal/builders/generic/attest.go:63]: Adding a simple test in attest_test.go might also be good.

**Review by ianlewis — APPROVED** [review]
Does the Go builder handle this ok? I guess it controls the subject's output location when it's compiled.

File: internal/builders/generic/attest.go (hunk 1 above)
Comment (asraa) [reply internal/builders/generic/attest.go:63]: done, thanks!

**Comment by asraa** [comment]
> Does the Go builder handle this ok? I guess it controls the subject's output location when it's compiled.

Good call, checking on this. It doesn't currently enforce that even when setting the output from the config (it checks if it's under the current working directory, but even a sub-directory would count). Double-checking more.

**Comment by ianlewis** [comment]
> > Does the Go builder handle this ok? I guess it controls the subject's output location when it's compiled.
> 
> Good call, checking on this. It doesn't currently enforce that even when setting the output from the config during dry run (it checks if it's under the current working directory, but even a sub-directory would count). Double-checking more.

If it's an issue I can follow up later in my day today.

**Comment by asraa** [comment]
OK, I don't think so. The config could end up setting
```
binary: ./subdirectory/binary-{{ .Os }}-{{ .Arch }}
```

and this would be allowed, and would fail on `CreateNewFileUnderCurrentDirectory`

**Comment by ianlewis** [comment]
> OK, I don't think so. The config could end up setting
> 
> ```
> binary: ./subdirectory/binary-{{ .Os }}-{{ .Arch }}
> ```
> 
> and this would be allowed, and would fail on `CreateNewFileUnderCurrentDirectory`

Yeah, lets go a separate PR for the Go builder.

File: internal/builders/generic/attest.go [hunk 2]
```diff
@@ -20,6 +20,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"os"
+	"path"
```
Comment (laurentsimon) [inline internal/builders/generic/attest.go:23]: why not set a default value for the provenance file in https://github.com/slsa-framework/slsa-github-generator/blob/main/.github/workflows/generator_generic_slsa3.yml#L37, and let users provide the name they want?

Comment (laurentsimon) [reply internal/builders/generic/attest.go:23]: I think we may run into problems with the secure-upload-artifact Actions. They currently assume files are in the root folder, I think

Comment (laurentsimon) [reply internal/builders/generic/attest.go:23]: edit: we use the upload-artifact, not the secure one so my last comment is wrong

**Review by laurentsimon — APPROVED** [review]

File: internal/builders/generic/attest.go (hunk 2 above)
Comment (ianlewis) [reply internal/builders/generic/attest.go:23]: > why not set a default value for the provenance file in https://github.com/slsa-framework/slsa-github-generator/blob/main/.github/workflows/generator_generic_slsa3.yml#L37?

I'd argue against this, if just for the fact that users would have to specify the input when the didn't before so it adds friction to upgrading and folks have complained previously about having to add new inputs.

**Comment by ianlewis** [comment]
> Let's amend the doc to say the files don't support folders. I have not seen anyone need this feature when using Goreleaser.

I'd say it's reasonable to support it if goreleaser supports it, but we can prioritize it when someone complains.

**Pull request merged**